core/index/index_writer.cpp

```diff
@@ -48,108 +48,129 @@ constexpr size_t kNonUpdateRecord = std::numeric_limits<size_t>::max();
   // ending doc_id to consider in 'segment.meta' (exclusive)
   const size_t doc_id_end_;
-  // doc_ids masked in segment_meta
-  document_mask docs_mask_;
   // copy so that it can be moved into 'index_writer::pending_state_'
-  index_meta::index_segment_t segment_;
+  IndexSegment segment_;
+  // doc_ids masked in SegmentMeta
+  document_mask docs_mask_;
   // modification contexts referenced by 'update_contexts_'
-  std::span<const index_writer::modification_context> modification_contexts_;
-  // update contexts for documents in segment_meta
+  std::span<const IndexWriter::ModificationContext> modification_contexts_;
+  // update contexts for documents in SegmentMeta
   std::span<const segment_writer::update_context> update_contexts_;
+  std::shared_ptr<const SegmentReaderImpl> reader_;
 
-  flush_segment_context(
-    const index_meta::index_segment_t& segment, size_t doc_id_begin,
-    size_t doc_id_end,
+  FlushSegmentContext(
+    std::shared_ptr<const SegmentReaderImpl>&& reader, IndexSegment&& segment,
+    document_mask&& docs_mask, size_t doc_id_begin, size_t doc_id_end,
     std::span<const segment_writer::update_context> update_contexts,
-    std::span<const index_writer::modification_context> modification_contexts)
-    : doc_id_begin_(doc_id_begin),
-      doc_id_end_(doc_id_end),
-      segment_(segment),
+    std::span<const IndexWriter::ModificationContext> modification_contexts)
+    : doc_id_begin_{doc_id_begin},
+      doc_id_end_{doc_id_end},
+      segment_{std::move(segment)},
+      docs_mask_{std::move(docs_mask)},
       modification_contexts_{modification_contexts},
-      update_contexts_{update_contexts} {
+      update_contexts_{update_contexts},
+      reader_{std::move(reader)} {
     IRS_ASSERT(doc_id_begin_ <= doc_id_end_);
     IRS_ASSERT(doc_id_end_ - doc_limits::min() <= segment_.meta.docs_count);
     IRS_ASSERT(update_contexts.size() == segment_.meta.docs_count);
   }
 };
 
-std::vector<index_file_refs::ref_t> extract_refs(
-  const ref_tracking_directory& dir) {
-  std::vector<index_file_refs::ref_t> refs;
-  // FIXME reserve
-
-  auto visitor = [&refs](const index_file_refs::ref_t& ref) {
-    refs.emplace_back(ref);
-    return true;
-  };
-  dir.visit_refs(visitor);
-
-  return refs;
-}
-
 // Apply any document removals based on filters in the segment.
 // modifications where to get document update_contexts from
 // docs_mask where to apply document removals to
 // readers readers by segment name
 // meta key used to get reader for the segment to evaluate
 // min_modification_generation smallest consider modification generation
 // Return if any new records were added (modification_queries_ modified).
-bool add_document_mask_modified_records(
-  std::span<index_writer::modification_context> modifications,
-  document_mask& docs_mask, readers_cache& readers, segment_meta& meta,
-  size_t min_modification_generation = 0) {
+void RemoveFromExistingSegment(
+  std::vector<doc_id_t>& deleted_docs,
```

Comment:
Is it necessary to have deleted_docs a out parameter instead of return variable?

Reply:
Yes, because we want to reuse memory.